skills-report-pre-cache-api-responses

anathomical opened the pull request:
## WHAT
- Refactor `classroom_scores_by_` both `skill` and `user` to use new `Compose` services
- Add a new `for_cache` flag to the `question_responses_for_classroom` and `_by_skill` run parameter `for_cache`
- Add individual workers for pre-caching both "by Student" and "by Skill" report data
- Add a fan-out worker to enqueue individual workers for each classroom of a specific user
- Enqueue the fan-out worker for a teacher when they load their dashboard if they are a premium user
## WHY
- We want to separate final payload processing from localized data fetches, ensuring that `Compose` services never fetch from database, so you can call it and know there won't be db-related performance costs
- The new `for_cache` flag is a "maximalist" version of the data.  It allows for a single data fetch of all facets of the appropriate data, intending to slice it up in Ruby
- Have workers that fetch all possible report data, and then slice it into its various facets in memory to construct and cache payloads with instead of fetching each facet individually from the database
- Data must be fetched on a per-classroom basis, so having the cache workers take the classroom_id as an argument makes sense, but we want a straightforward way of caching report data for all of a given teacher's classrooms
- We want to actually pre-cache reporting data for most teachers so that when they reach the reports, performance is super-snappy
## HOW
- Create new services to `Compose` data, and insert it into the `ClassroomScore`s services
- Add the flag to `initialize`, tweak the query to remove many normal restrictions when the flag is passed, and add some additional values to the `Response` objects to allow for in-Ruby filtering in lieu of in-Postgres filtering
- Create workers that take `user_id` and `classroom_id`, fetch the data, slice it up, and use `classroom_cache` to cache this data in the same keys the controllers fetch from
- Add a worker that uses `classrooms_i_teach` to fan out for all classrooms of a given user
- Add a `before_action` to enqueue the fan-out worker for a teacher if they are a premium user

### Notion Card Links
https://www.notion.so/quill/Skills-Report-2-0-190d42e6f941801bbab8e37e97e68147

### What have you done to QA this feature?
- Manually clear cache
- Load teacher's dashboard
- Navigate to Skills Report
- Confirm that selection load times are < 1 second
- Confirm that new database calls are not being made for the report
- Use a script to confirm that the value being cached matches the value that is generated by the narrower database-based data generation

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes

anathomical committed:
Update the Skills service chain in prep for pre-caching shims

anathomical committed:
Establish framework to trigger pre-caching behavior on dashboard load

anathomical committed:
Add ability to fetch "for_cache" which is the maximal data-set

This is to support parsing through this data in Ruby and caching it all in separate keys to support pre-caching

anathomical committed:
Add worker to pre-cache Skills Report data for specific teachers

anathomical committed:
Only trigger Skills Report caching for premium users

anathomical committed:
Fix linting

anathomical committed:
Refactor to use new Compose service to prep for pre-caching this data

anathomical committed:
Update data fetching logic to handle maximalist for_cache flag

anathomical committed:
Add worker to pre-cache "by Student" report views

anathomical committed:
Add spec for new worker

anathomical committed:
Fix linting

anathomical committed:
Move cache key into constant

happythenewsad commented on services/QuillLMS/app/workers/skills_report/cache_users_api_data_worker.rb:
```diff
@@ -0,0 +1,69 @@
+    private def user_hash = @user_hash ||= User.where(id: user_ids).pluck(:id, :name).to_h
+    private def user_ids = all_rows.map(&:user_id).uniq
+
+    private def filter_classroom_history(rows, classroom_id) = rows.filter { |row| row.classroom_id == classroom_id }
```
This method and several others are duplicates of those in `cache_skills_api_data_worker`

anathomical replied:
I think that this is a good call-out, but will address DRY issues in a separate PR to keep this simple.

happythenewsad commented on services/QuillLMS/app/services/skills_report/question_responses_for_classroom_by_skill.rb:
```diff
@@ -51,12 +51,20 @@ class QuestionResponsesForClassroomBySkill < ApplicationService
         FROM students_classrooms sc
             JOIN activity_sessions act_s ON sc.student_id = act_s.user_id
-                #{'AND act_s.classroom_unit_id IN (SELECT id FROM classroom_unit_ids)' unless all_history}
+                #{'AND act_s.classroom_unit_id IN (SELECT id FROM classroom_unit_ids)' unless all_history || for_cache}
```
Optional style thing: I find SQL is more readable and maintainable when Ruby logic is maximally extracted. How about:
```ruby
def join_classroom_unit_ids 
  return if all_history || for_cache
  'AND act_s.classroom_unit_id IN (SELECT id FROM classroom_unit_ids)'
end
```

happythenewsad reviewed (approved):
There's an opportunity to DRY up some methods, but overall this is very clearly organized . Nice... composition

anathomical committed:
Try using fetch-multi to see if it gets around rate limiting

anathomical committed:
Try writing no matter what, and reduce load on ELL flag

anathomical committed:
Use correct syntax for write_multi

anathomical committed:
Test not pre-caching student selections

anathomical committed:
Move WHERE condition earlier in the chain to speed query up

By doing the WHERE before any JOINS, we speed up the JOINs themselves

anathomical committed:
WIP updating specs for new logic

anathomical committed:
Only pre-cache the views a teacher can land on the first time

So any toggle that can be persisted in localStorage

anathomical committed:
Try sorting classrooms by name before enqueuing for pre-cache

anathomical committed:
Tweak query so that it works with studentless classrooms

anathomical committed:
Testing new high-performance query for Skills

anathomical committed:
Test high-performance rewrite of by-student

anathomical committed:
Fix some linting

anathomical committed:
Update "By Student" APIs to provide placeholders like "By Skill"

Instead of relying on the database lookup to find them

anathomical committed:
Update user pre-caching to pre-cache appropriate user placeholders

anathomical committed:
Try changing EMA Trend threshold to change class-level aggregation

anathomical committed:
Update formatTrend to not always select "Steady"

anathomical committed:
Use LEFT OUTER JOIN for questions

Because Proofreader activities, and if we ever add them, Evidence activities, do not have Question records, but we still want to include them

anathomical committed:
Re-set the threshold percentage to 5

anathomical committed:
Fix "by skills" average trend bug

dandrabik commented on services/QuillLMS/app/controllers/teachers/classroom_manager_controller.rb:
```diff
@@ -391,7 +391,9 @@ def set_diagnostic_variables
-  private def cache_diagnostic_insights_recommendation_count
+  private def pre_cache_report_data
     DiagnosticInsights::CacheRecommendationCountsWorker.perform_async(current_user.id)
+    # Only pre-cache Skills Report if the user is premium since it's a premium-only report
```
Consider an early `return unless current_user.subscription` here. That might make the comment less needed.

dandrabik commented on services/QuillLMS/app/services/skills_report/classroom_scores_by_skill.rb:
```diff
@@ -13,45 +13,21 @@ def initialize(classroom_id:, concept_id: nil, user_id: nil, all_history: false,
-          id = skill_data[:id]
-          SkillResponsesProcessor.run(user_responses:, name: skill_name, id:, questions_hash:, total_classroom_students:)
-        end
+      return ComposeScoresBySkill.run(concepts:, question_responses:, questions_hash:, total_classroom_students:)
```
You can remove the `return`

dandrabik commented on services/QuillLMS/app/services/skills_report/compose_scores_by_user.rb:
```diff
@@ -0,0 +1,21 @@
+
+    def run
+      response_hash
+        .to_a
```
You don't need this `to_a`. `hash.map { |user_id, response| ... }`

dandrabik commented on services/QuillLMS/app/services/skills_report/question_responses_for_classroom.rb:
```diff
@@ -53,27 +62,39 @@ class QuestionResponsesForClassroom < ApplicationService
-    private def sql = ActiveRecord::Base.sanitize_sql([query_sql, { classroom_id:, concept_id: }])
+    private def sql = ActiveRecord::Base.sanitize_sql([query_sql, { classroom_id:, concept_ids: }])
+
+    private def concept_ids
```
Take or leave, consider making this its own method for readability:

```ruby
private def all_skills_report_concept_ids = Concept.in_skills_report.visible.pluck(:id)

...
return all_skills_report_concept_ids if for_cache

[concept_id]

anathomical committed:
Update specs to handle newer code updates

dandrabik reviewed (approved):
Overall this looks really good. Well structured and easy to read. Small take or leave notes, but looks great! :ship:

anathomical committed:
Include archived activity sessions in reporting data

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into skills-report-pre-cache-api-responses

anathomical committed:
Update test for new trend-display logic

anathomical committed:
Clean up pre-cache logic to be clearer and more readable

anathomical committed:
Refactor the service to extract string-interpolation

Put that into its own methods so that the queries read cleaner

anathomical committed:
Tweaks based on PR feedback
